File: app/src/main/java/sugtao4423/yandereviewer/PostAdapter.java
package sugtao4423.yandereviewer;

import android.content.Context;
import android.graphics.Color;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.loopj.android.image.SmartImageView;

import java.util.ArrayList;
import java.util.HashMap;

import yandere4j.data.Post;

public class PostAdapter extends RecyclerView.Adapter<PostAdapter.ViewHolder>{

    private MainActivity mainActivity;
    private long readedId = -1L;
    private ArrayList<Post> data;
    private LayoutInflater inflater;
    private HashMap<Post, Integer> multiSelectedItems;
    private int cardViewBGColor;

    public PostAdapter(MainActivity mainActivity){
        this.mainActivity = mainActivity;
        data = new ArrayList<Post>();
        inflater = (LayoutInflater)mainActivity.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        readedId = PreferenceManager.getDefaultSharedPreferences(mainActivity.getApplicationContext()).getLong(Keys.READEDID, -1L);
        multiSelectedItems = new HashMap<Post, Integer>();
        cardViewBGColor = new CardView(mainActivity).getCardBackgroundColor().getDefaultColor();
    }

    @Override
    public PostAdapter.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int position){
        return new ViewHolder(inflater.inflate(R.layout.grid_item_layout, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position){
        if(!(data != null && data.size() > position && data.get(position) != null))
            return;
        final Post item = data.get(position);

        holder.itemView.setOnClickListener(mainActivity.getOnCardClickListener(item));
        holder.itemView.setOnLongClickListener(mainActivity.getOnCardLongClickListener());

        if(item.getMD5().equals("LOADMORE")){
            holder.image.setImageResource(R.drawable.plus);
            holder.imageSize.setText("Load More");
            holder.imageSize.setBackgroundColor(Color.parseColor("#ffffff"));
            return;
        }
        holder.image.setImageUrl(item.getPreview().getUrl(), null, R.drawable.ic_action_refresh);
        holder.imageSize.setText(item.getFile().getWidth() + "x" + item.getFile().getHeight());
        if(readedId < item.getId())
            holder.imageSize.setBackgroundColor(Color.parseColor("#e1bee7"));
        else
            holder.imageSize.setBackgroundColor(Color.parseColor("#ffffff"));
        if(multiSelectedItems.keySet().contains(item))
            holder.itemView.setBackgroundColor(Color.parseColor("#B3E5FC"));
        else
            holder.itemView.setBackgroundColor(cardViewBGColor);
    }

    @Override
    public int getItemCount(){
        if(data != null)
            return data.size();
        else
            return 0;
    }

    public void clear(){
        int size = data.size();
        data.clear();
        notifyItemRangeRemoved(0, size);
    }

    public void add(Post post){
        data.add(post);
        notifyItemInserted(data.size() - 1);
    }

    public void addAll(Post[] posts){
        int pos = data.size();
        for(Post p : posts)
            data.add(p);
        notifyItemRangeInserted(pos, posts.length);
    }

    public void remove(Post post){
        int pos = data.indexOf(post);
        data.remove(post);
        notifyItemRemoved(pos);
    }

    public boolean isPostSelected(Post post){
        return multiSelectedItems.keySet().contains(post);
    }

    public void setPostSelected(Post post, boolean isSelect){
        int pos = data.indexOf(post);
        if(isSelect)
            multiSelectedItems.put(post, pos);
        else
            multiSelectedItems.remove(post);
        notifyItemChanged(pos);
    }

    public Post[] getSelectedPosts(){
        return multiSelectedItems.keySet().toArray(new Post[multiSelectedItems.size()]);
    }

    public void clearSelectedPosts(){
        Integer[] destroySelectedPostPos = multiSelectedItems.values().toArray(new Integer[multiSelectedItems.size()]);
        multiSelectedItems.clear();
        for(int i : destroySelectedPostPos)
            notifyItemChanged(i);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder{
        SmartImageView image;
        TextView imageSize;

        public ViewHolder(View v){
            super(v);
            image = (SmartImageView)v.findViewById(R.id.postImage);
            imageSize = (TextView)v.findViewById(R.id.imageSize);
        }
    }

}

File: app/src/main/java/sugtao4423/yandereviewer/PostAdapter.kt
package sugtao4423.yandereviewer

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.LazyHeaders
import yandere4j.Post
import yandere4j.Yandere4j

class PostAdapter(private val mainActivity: MainActivity) : RecyclerView.Adapter<PostAdapter.ViewHolder>() {

    private var readedId = PreferenceManager.getDefaultSharedPreferences(mainActivity.applicationContext).getLong(Keys.READEDID, -1L)
    private val data = ArrayList<Post>()
    private val inflater = mainActivity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    private val multiSelectedItems = HashMap<Post, Int>()
    private var cardViewBGColor = CardView(mainActivity).cardBackgroundColor.defaultColor

    override fun onCreateViewHolder(viewGroup: ViewGroup, position: Int): ViewHolder {
        return ViewHolder(inflater.inflate(R.layout.grid_item_layout, viewGroup, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (data.size <= position) {
            return
        }
        val item = data[position]

        holder.itemView.setOnClickListener(mainActivity.getOnCardClickListener(item))
        holder.itemView.setOnLongClickListener(mainActivity.getOnCardLongClickListener())

        val glideHeader = LazyHeaders.Builder().addHeader("User-Agent", Yandere4j.USER_AGENT).build()
        val glideUrl = GlideUrl(item.preview.url, glideHeader)
        Glide.with(mainActivity).load(glideUrl).placeholder(R.drawable.ic_action_refresh).into(holder.image)
        holder.imageSize.text = "${item.file.width}x${item.file.height}"

        holder.imageSize.setBackgroundColor(
                if (readedId < item.id) {
                    Color.parseColor("#e1bee7")
                } else {
                    Color.parseColor("#ffffff")
                }
        )

        holder.itemView.setBackgroundColor(
                if (multiSelectedItems.keys.contains(item)) {
                    Color.parseColor("#B3E5FC")
                } else {
                    cardViewBGColor
                }
        )
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun clear() {
        val size = data.size
        data.clear()
        notifyItemRangeRemoved(0, size)
    }

    fun add(post: Post) {
        data.add(post)
        notifyItemInserted(data.size - 1)
    }

    fun addAll(posts: Array<Post>) {
        val pos = data.size
        posts.map {
            data.add(it)
        }
        notifyItemRangeInserted(pos, posts.size)
    }

    fun remove(post: Post) {
        val pos = data.indexOf(post)
        data.remove(post)
        notifyItemRemoved(pos)
    }

    fun isPostSelected(post: Post): Boolean {
        return multiSelectedItems.keys.contains(post)
    }

    fun setPostSelected(post: Post, isSelect: Boolean) {
        val pos = data.indexOf(post)
        if (isSelect) {
            multiSelectedItems[post] = pos
        } else {
            multiSelectedItems.remove(post)
        }
        notifyItemChanged(pos)
    }

    fun getSelectedPosts(): Array<Post> {
        return multiSelectedItems.keys.toTypedArray()
    }

    fun clearSelectedPosts() {
        val destroySelectedPostPos = multiSelectedItems.values.toTypedArray()
        multiSelectedItems.clear()
        destroySelectedPostPos.map {
            notifyItemChanged(it)
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val image: ImageView = itemView.findViewById(R.id.postImage)
        val imageSize: TextView = itemView.findViewById(R.id.imageSize)
    }

}